### DISCUSSION TITLE
--mux-yaml not available in avocado 8.7

### DISCUSSION THREAD

**Discussion by marioaag** [avocado-framework/avocado#4545] [Q&A] [answered]
Try to use the switch --mux-yaml to use info coming from a yaml file in test.py.data dir but it complains about it:

avocado run: error: unrecognized arguments: --mux-yaml

I looked around in the documentation but didn't find the way to install or enable this plugin

Any suggestion?

Thanks in advance

**Comment by willianrampazzo** [comment]
Could you paste the complete command and the output? The `avocado plugins` command shows all the plugins you have available. If possible, add it here too.

**Reply by marioaag** [reply]
marioalvarado@MacBook-Pro testcases % avocado run --mux-yaml check_host_status.py.data/check_host_status.yaml check_host_status.py 
usage: avocado run [-h] [-p NAME_VALUE] [--test-runner RUN.TEST_RUNNER] [-d]
                   [--dry-run-no-cleanup] [--force-job-id RUN.UNIQUE_JOB_ID]
                   [--job-results-dir DIRECTORY] [--job-category CATEGORY]
                   [--job-timeout SECONDS] [--failfast] [--keep-tmp]
                   [--ignore-missing-references] [--disable-sysinfo]
                   [--execution-order RUN.EXECUTION_ORDER]
                   [--store-logging-stream STREAM[:LEVEL] [STREAM[:LEVEL]
                   ...]] [--log-test-data-directories]
                   [--output-check-record {none,stdout,stderr,both,combined,all}]
                   [--disable-output-check]
                   [--loaders RUN.LOADERS [RUN.LOADERS ...]]
                   [--external-runner RUN.EXTERNAL_RUNNER]
                   [--external-runner-chdir {runner,test}]
                   [--external-runner-testdir DIRECTORY] [-t TAGS]
                   [--filter-by-tags-include-empty]
                   [--filter-by-tags-include-empty-key] [--html FILE]
                   [--open-browser] [--disable-html-job-result] [--journal]
                   [--json FILE] [--disable-json-job-result]
                   [--json-variants-load JSON.VARIANTS.LOAD]
                   [--nrunner-shuffle]
                   [--nrunner-status-server-listen NRUNNER.STATUS_SERVER_LISTEN]
                   [--nrunner-status-server-uri NRUNNER.STATUS_SERVER_URI]
                   [--nrunner-max-parallel-tasks NRUNNER.MAX_PARALLEL_TASKS]
                   [--nrunner-spawner NRUNNER.SPAWNER]
                   [--spawner-podman-bin SPAWNER.PODMAN.BIN]
                   [--spawner-podman-image SPAWNER.PODMAN.IMAGE]
                   [--replay RUN.REPLAY.JOB_ID]
                   [--replay-test-status RUN.REPLAY.TEST_STATUS]
                   [--replay-ignore RUN.REPLAY.IGNORE] [--replay-resume]
                   [--tap FILE] [--disable-tap-job-result]
                   [--tap-include-logs] [--wrapper SCRIPT[:EXECUTABLE]]
                   [--xunit FILE] [--disable-xunit-job-result]
                   [--xunit-job-name JOB.RUN.RESULT.XUNIT.JOB_NAME]
                   [--xunit-max-test-log-chars SIZE] [-z]
                   [TEST_REFERENCE [TEST_REFERENCE ...]]

positional arguments:
  TEST_REFERENCE        List of test references (aliases or paths)

optional arguments:
  -h, --help            show this help message and exit
  -p NAME_VALUE, --test-parameter NAME_VALUE
                        Parameter name and value to pass to all tests. This is
                        only applicable when not using a varianter plugin.
                        This option format must be given in the NAME=VALUE
                        format, and may be given any number of times, or per
                        parameter.
  --test-runner RUN.TEST_RUNNER
                        Selects the runner implementation from one of the
                        installed and active implementations. You can run
                        "avocado plugins" and find the list of valid runners
                        under the "Plugins that run test suites on a job
                        (runners) section. Defaults to "runner", which is the
                        conventional and traditional runner.
  -d, --dry-run         Instead of running the test only list them and log
                        their params.
  --dry-run-no-cleanup  Do not automatically clean up temporary directories
                        used by dry-run
  --force-job-id RUN.UNIQUE_JOB_ID
                        Forces the use of a particular job ID. Used internally
                        when interacting with an avocado server. You should
                        not use this option unless you know exactly what
                        you're doing
  --job-results-dir DIRECTORY
                        Forces to use of an alternate job results directory.
  --job-category CATEGORY
                        Categorizes this within a directory with the same
                        name, by creating a link to the job result directory
  --job-timeout SECONDS
                        Set the maximum amount of time (in SECONDS) that tests
                        are allowed to execute. Values <= zero means "no
                        timeout". You can also use suffixes, like: s
                        (seconds), m (minutes), h (hours).
  --failfast            Enable the job interruption on first failed test.
  --keep-tmp            Keep job temporary files (useful for avocado
                        debugging).
  --ignore-missing-references
                        Force the job execution, even if some of the test
                        references are not resolved to tests. "on" and "off"
                        will be deprecated soon.
  --disable-sysinfo     Enable or disable sysinfo information. Like hardware
                        details, profiles, etc.
  --execution-order RUN.EXECUTION_ORDER
                        Defines the order of iterating through test suite and
                        test variants
  --store-logging-stream STREAM[:LEVEL] [STREAM[:LEVEL] ...]
                        Store given logging STREAMs in
                        "$JOB_RESULTS_DIR/$STREAM.$LEVEL."
  --log-test-data-directories
                        Logs the possible data directories for each test. This
                        is helpful when writing new tests and not being sure
                        where to put data files. Look for "Test data
                        directories" in your test log
  --html FILE           Enable HTML output to the FILE where the result should
                        be written. The value - (output to stdout) is not
                        supported since not all HTML resources can be embedded
                        into a single file (page resources will be copied to
                        the output file dir)
  --open-browser        Open the generated report on your preferred browser.
                        This works even if --html was not explicitly passed,
                        since an HTML report is always generated on the job
                        results dir.
  --disable-html-job-result
                        Enables default HTML result in the job results
                        directory. File will be named "results.html".
  --journal             Records test status changes (for use with avocado-
                        journal-replay and avocado-server)
  --json FILE           Enable JSON result format and write it to FILE. Use
                        "-" to redirect to the standard output.
  --disable-json-job-result
                        Enables default JSON result in the job results
                        directory. File will be named "results.json".
  --tap FILE            Enable TAP result output and write it to FILE. Use "-"
                        to redirect to standard output.
  --disable-tap-job-result
                        Enables default TAP result in the job results
                        directory. File will be named "results.tap"
  --tap-include-logs    Include test logs as comments in TAP output.
  -z, --archive         Archive (ZIP) files generated by tests

output and result format:
  --xunit FILE          Enable xUnit result format and write it to FILE. Use
                        "-" to redirect to the standard output.
  --disable-xunit-job-result
                        Enables default xUnit result in the job results
                        directory. File will be named "results.xml".
  --xunit-job-name JOB.RUN.RESULT.XUNIT.JOB_NAME
                        Override the reported job name. By default uses the
                        Avocado job name which is always unique. This is
                        useful for reporting in Jenkins as it only evaluates
                        first-failure from jobs of the same name.
  --xunit-max-test-log-chars SIZE
                        Limit the attached job log to given number of
                        characters (k/m/g suffix allowed)

output check arguments:
  --output-check-record {none,stdout,stderr,both,combined,all}
                        Record the output produced by each test (from stdout
                        and stderr) into both the current executing result and
                        into reference files. Reference files are used on
                        subsequent runs to determine if the test produced the
                        expected output or not, and the current executing
                        result is used to check against a previously recorded
                        reference file. Valid values: "none" (to explicitly
                        disable all recording) "stdout" (to record standard
                        output *only*), "stderr" (to record standard error
                        *only*), "both" (to record standard output and error
                        in separate files), "combined" (for standard output
                        and error in a single file). "all" is also a valid but
                        deprecated option that is a synonym of "both".
  --disable-output-check
                        Disables test output (stdout/stderr) check. If this
                        option is given, no output will be checked, even if
                        there are reference files present for the test.

loader options:
  --loaders RUN.LOADERS [RUN.LOADERS ...]
                        Overrides the priority of the test loaders. You can
                        specify either @loader_name or TEST_TYPE. By default
                        it tries all available loaders according to priority
                        set in settings->plugins.loaders.
  --external-runner RUN.EXTERNAL_RUNNER
                        Path to an specific test runner that allows the use of
                        its own tests. This should be used for running tests
                        that do not conform to Avocado's SIMPLE test interface
                        and can not run standalone. Note: the use of
                        --external-runner overwrites the --loaders to
                        'external_runner'
  --external-runner-chdir {runner,test}
                        Change directory before executing tests. This option
                        may be necessary because of requirements and/or
                        limitations of the external test runner. If the
                        external runner requires to be run from its own base
                        directory, use 'runner' here. If the external runner
                        runs tests based on files and requires to be run from
                        the directory where those files are located, use
                        'test' here and specify the test directory with the
                        option '--external-runner-testdir'.
  --external-runner-testdir DIRECTORY
                        Where test files understood by the external test
                        runner are located in the filesystem. Obviously this
                        assumes and only applies to external test runners that
                        run tests from files

filtering parameters:
  -t TAGS, --filter-by-tags TAGS
                        Filter tests based on tags
  --filter-by-tags-include-empty
                        Include all tests without tags during filtering. This
                        effectively means they will be kept in the test suite
                        found previously to filtering.
  --filter-by-tags-include-empty-key
                        Include all tests that do not have a matching key in
                        its key:val tags. This effectively means those tests
                        will be kept in the test suite found previously to
                        filtering.

JSON serialized based varianter options:
  --json-variants-load JSON.VARIANTS.LOAD
                        Load the Variants from a JSON serialized file

nrunner specific options:
  --nrunner-shuffle     Shuffle the tasks to be executed
  --nrunner-status-server-listen NRUNNER.STATUS_SERVER_LISTEN
                        URI for listing the status server. Usually a
                        "HOST:PORT" string
  --nrunner-status-server-uri NRUNNER.STATUS_SERVER_URI
                        URI for connecting to the status server, usually a
                        "HOST:PORT" string. Use this if your status server is
                        in another host, or different port
  --nrunner-max-parallel-tasks NRUNNER.MAX_PARALLEL_TASKS
                        Number of maximum number tasks running in parallel.
                        You can disable parallel execution by setting this to
                        1. Defaults to the amount of CPUs on this machine.
  --nrunner-spawner NRUNNER.SPAWNER
                        Spawn tasks in a specific spawner. Available spawners:
                        'process' and 'podman'

podman spawner specific options:
  --spawner-podman-bin SPAWNER.PODMAN.BIN
                        Path to the podman binary
  --spawner-podman-image SPAWNER.PODMAN.IMAGE
                        Image name to use when creating the container

job replay:
  --replay RUN.REPLAY.JOB_ID
                        Replay a job identified by its (partial) hash id. Use
                        "--replay" latest to replay the latest job.
  --replay-test-status RUN.REPLAY.TEST_STATUS
                        Filter tests to replay by test status.
  --replay-ignore RUN.REPLAY.IGNORE
                        Ignore variants and/or configuration from the source
                        job.
  --replay-resume       Resume an interrupted job

wrapper support:
  --wrapper SCRIPT[:EXECUTABLE]
                        Use a script to wrap executables run by a test. The
                        wrapper is either a path to a script (AKA a global
                        wrapper) or a path to a script followed by colon
                        symbol (:), plus a shell like glob to the target
                        EXECUTABLE. Multiple wrapper options are allowed, but
                        only one global wrapper can be defined.

avocado run: error: unrecognized arguments: --mux-yaml
Perhaps a plugin is missing; run 'avocado plugins' to list the installed ones

**Reply by clebergnu** [reply]
@marioaag you need to install the plugin.  If you're using `pip`, this is the package: https://pypi.org/project/avocado-framework-plugin-varianter-yaml-to-mux/

**Reply by marioaag** [reply]
Thanks @clebergnu, this is what I was looking for but trying to install, I'm getting the next issue:
``` marioalvarado@MacBook-Pro testcases % python3 -m pip install https://pypi.org/project/avocado-framework-plugin-varianter-yaml-to-mux/ 
Defaulting to user installation because normal site-packages is not writeable
Collecting https://pypi.org/project/avocado-framework-plugin-varianter-yaml-to-mux/
  Downloading https://pypi.org/project/avocado-framework-plugin-varianter-yaml-to-mux/ (79 kB)
     |████████████████████████████████| 79 kB 487 kB/s 
  ERROR: Cannot unpack file /private/var/folders/pn/xdk10w_12dx8_bvjjrfh35hh0000gn/T/pip-unpack-77z20an5/avocado-framework-plugin-varianter-yaml-to-mux (downloaded from /private/var/folders/pn/xdk10w_12dx8_bvjjrfh35hh0000gn/T/pip-req-build-xa79d8u4, content-type: text/html; charset=UTF-8); cannot detect archive format
ERROR: Cannot determine archive format of /private/var/folders/pn/xdk10w_12dx8_bvjjrfh35hh0000gn/T/pip-req-build-xa79d8u4 ```

**Reply by willianrampazzo** [reply]
Try `pip install avocado-framework-plugin-varianter-yaml-to-mux`. Use `--user` if you are not installing the package system-wide.

**Reply by marioaag** [reply]
Thanks that worked!
Is this functionality already implemented for nrunner?
What is the easiest way to define parameters and tests to be run using a yaml file? something with an structure like this:
```
common: !mux
    parameter1: 1
    parameter2: 1
tests: !mux
    test1:
        parameter3: 3
        parameter4: 4
    test2:
        parameter3: 5
        parameter4: 6
```

**Comment by marioaag** [comment]
@willianrampazzo, is the varianter plugging already implemented for the nrunner?

**Reply by willianrampazzo** [reply, marked as answer]
Unfortunately, it is not available on nrunner yet.

**Reply by marioaag** [reply]
No problem, just wanted to be sure, is there an ETA for it? if not what should be the steeps to get involve and help with its implementation?
